### PR TITLE
perf: Parallelize calculation of deep codeword inner product

### PR DESCRIPTION
**Pull request by Sword-Smith**
This speeds up the step of calculating the combined deep codeword from 900ms to 370ms for the reference parameters expansion factor 4, padded height 2^20.

There might be even faster parallelization options for this calculation but this is already a decent speedup of a factor 2.4

### CODE REVIEW

File: triton-vm/src/stark.rs [hunk 1]
```diff
@@ -342,8 +342,15 @@ impl Stark {
...
         )
         .unwrap();
-        let weighted_deep_codeword_components = &deep_codeword_components * &weights.deep;
-        let deep_codeword = weighted_deep_codeword_components.sum_axis(Axis(1));
+        let mut deep_codeword = Vec::with_capacity(short_domain.length);
+        deep_codeword_components
+            .axis_iter(Axis(0))
+            .into_par_iter()
+            .map(|chunk| {
+                chunk[0] * weights.deep[0] + chunk[1] * weights.deep[1] + chunk[2] * weights.deep[2]
```
Comment (aszepieniec) [inline triton-vm/src/stark.rs:350]: Why call the variable `chunk`? I don't think it spans more than 1 row. Maybe I am misunderstanding something ...

Comment (Sword-Smith) [reply triton-vm/src/stark.rs:350]: Agreed. Maybe a better name is column? Or segment?

File: triton-vm/src/stark.rs [hunk 2]
```diff
@@ -332,25 +333,30 @@ impl Stark {
...
-        let deep_codeword_components = Array2::from_shape_vec(
-            [short_domain.length, NUM_DEEP_CODEWORD_COMPONENTS].f(),
-            deep_codeword_components.concat(),
-        )
-        .unwrap();
-        let weighted_deep_codeword_components = &deep_codeword_components * &weights.deep;
-        let deep_codeword = weighted_deep_codeword_components.sum_axis(Axis(1));
+        ]
+        .into_par_iter()
+        .zip_eq(weights.deep.to_vec().into_par_iter())
```
Comment (jan-ferdinand) [inline triton-vm/src/stark.rs:342]:
```suggestion
        .zip_eq(weights.deep.to_vec())
```
Thanks to the trait bounds of `.zip_eq()` (outlined below), the `.into_par_iter()` will be called automatically.

```rust
// rayon::iter::IndexedParallelIterator
fn zip_eq<Z>(self, zip_op: Z) -> ZipEq<Self, Z::Iter>
where
	Z: IntoParallelIterator,
	Z::Iter: IndexedParallelIterator,
```

File: triton-vm/src/stark.rs [hunk 3]
```diff
@@ -332,25 +333,30 @@ impl Stark {
...
+        .map(|(code_word, weight)| {
+            code_word
```
Comment (jan-ferdinand) [inline triton-vm/src/stark.rs:344]:
```suggestion
            codeword
```

This spelling is more consistent.

File: triton-vm/src/stark.rs [hunk 4]
```diff
@@ -332,25 +333,30 @@ impl Stark {
...
+                .into_par_iter()
+                .map(|cw_elem| cw_elem * weight)
+                .collect::<Vec<_>>()
```
Comment (jan-ferdinand) [inline triton-vm/src/stark.rs:347]:
```suggestion
                .collect()
```

This type is automatically inferred.

File: triton-vm/src/stark.rs [hunk 5]
```diff
@@ -332,25 +333,30 @@ impl Stark {
...
-        .unwrap();
-        let weighted_deep_codeword_components = &deep_codeword_components * &weights.deep;
-        let deep_codeword = weighted_deep_codeword_components.sum_axis(Axis(1));
+        ]
+        .into_par_iter()
+        .zip_eq(weights.deep.to_vec().into_par_iter())
+        .map(|(code_word, weight)| {
+            code_word
+                .into_par_iter()
+                .map(|cw_elem| cw_elem * weight)
```
Comment (jan-ferdinand) [inline triton-vm/src/stark.rs:346]:
```suggestion
                .map(|c| c * weight)
```

Shortening this name makes the double-closure a one liner. :tada:

File: triton-vm/src/stark.rs [hunk 6]
```diff
@@ -332,25 +333,30 @@ impl Stark {
...
+                .collect::<Vec<_>>()
+        })
+        .reduce(
+            || vec![XFieldElement::zero(); short_domain.length],
+            |a, b| a.into_iter().zip(b).map(|(c, d)| c + d).collect::<Vec<_>>(),
```
Comment (jan-ferdinand) [inline triton-vm/src/stark.rs:351]:
```suggestion
            |a, b| a.into_iter().zip(b).map(|(c, d)| c + d).collect(),
```

This type is automatically inferred.

**Review by jan-ferdinand — APPROVED** [review]
Thanks for the speedup! :fast_forward: I'm picking nits in the comments. I'm more than happy to apply all the nit-picky changes myself and merge the PR.

**Pull request merged**